File: src-tauri/src/events.rs
use serde::Serialize;

pub const APP_LOADED: &str = "APP_LOADED";

// DOWNLOAD
pub const DOWNLOAD_FILE_ADDED: &str = "DOWNLOAD_FILE_ADDED";
pub const DOWNLOAD_FILE_PROGRESS: &str = "DOWNLOAD_FILE_PROGRESS";
pub const DOWNLOAD_FILE_COMPLETED: &str = "DOWNLOAD_FILE_COMPLETED";
pub const DOWNLOAD_ALL_COMPLETE: &str = "DOWNLOAD_ALL_COMPLETE";
pub const DOWNLOAD_FILE_ERROR: &str = "DOWNLOAD_FILE_ERROR";
pub const DOWNLOAD_FILE_ABORTED: &str = "DOWNLOAD_FILE_ABORTED";
pub const CANCEL_DOWNLOAD: &str = "CANCEL_DOWNLOAD";

#[derive(Debug, Clone, Serialize)]
pub struct DownloadFileAdded {
    pub name: String,
    pub icon: String,
    pub size: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct DownloadFileProgress {
    pub name: String,
    pub progress: f32,
    pub speed: f32,
}

#[derive(Debug, Clone, Serialize)]
pub struct DownloadFileCompleted {
    pub name: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct DownloadFileAborted {
    pub name: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct DownloadFileError {
    pub name: String,
    pub error: String,
}

// UPLOAD
pub const UPLOAD_FILE_ADDED: &str = "UPLOAD_FILE_ADDED";
pub const UPLOAD_FILE_PROGRESS: &str = "UPLOAD_FILE_PROGRESS";
pub const UPLOAD_FILE_COMPLETED: &str = "UPLOAD_FILE_COMPLETED";
pub const UPLOAD_FILE_REMOVED: &str = "UPLOAD_FILE_REMOVED";
#[allow(unused)]
pub const UPLOAD_FILE_ERROR: &str = "UPLOAD_FILE_ERROR";

#[derive(Debug, Clone, Serialize)]
pub struct UploadFileAdded {
    pub name: String,
    pub icon: String,
    pub path: String,
    pub size: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct UploadFileProgress {
    pub path: String,
    pub progress: f32,
}

#[derive(Debug, Clone, Serialize)]
pub struct UploadFileRemoved {
    pub name: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct UploadFileCompleted {
    pub name: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct UploadFileError {
    pub name: String,
    pub error: String,
}

// REMOVE_FILE

File: src/lib/tauri/events.ts
// DOWNLOAD
//
export const DOWNLOAD_FILE_ADDED = 'DOWNLOAD_FILE_ADDED'
export const DOWNLOAD_FILE_PROGRESS = 'DOWNLOAD_FILE_PROGRESS'
export const DOWNLOAD_FILE_COMPLETED = 'DOWNLOAD_FILE_COMPLETED'
export const DOWNLOAD_ALL_COMPLETE = 'DOWNLOAD_ALL_COMPLETE'
export const DOWNLOAD_FILE_ERROR = 'DOWNLOAD_FILE_ERROR'
export const DOWNLOAD_FILE_ABORTED = 'DOWNLOAD_FILE_ABORTED'
export const CANCEL_DOWNLOAD = 'CANCEL_DOWNLOAD'

export type DownloadFileAdded = {
  name: string
  icon: string
  size: number
}
export type DownloadFileProgress = {
  name: string
  progress: number
  speed: number // bytes per microsecond
}
export type DownloadFileCompleted = { name: string; path: string }

export type DownloadFileAborted = {
  name: string
  reason: string
}

export type DownloadFileError = {
  name: string
  error: string
}

// UPLOAD
export const UPLOAD_FILE_ADDED = 'UPLOAD_FILE_ADDED'
export const UPLOAD_FILE_PROGRESS = 'UPLOAD_FILE_PROGRESS'
export const UPLOAD_FILE_COMPLETED = 'UPLOAD_FILE_COMPLETED'
export const UPLOAD_FILE_REMOVED = 'UPLOAD_FILE_REMOVED'
export const UPLOAD_FILE_ERROR = 'UPLOAD_FILE_ERROR'

export type UploadFileAdded = {
  name: string
  icon: string
  path: string
  size: number
}
export type UploadFileProgress = {
  path: string
  progress: number
}
export type UploadFileCompleted = { name: string }
export type UploadFileRemoved = { name: string }
export type UploadFileError = {
  name: string
  error: string
}
